src/lib.rs

```diff
@@ -32,7 +34,12 @@ use crate::panic::PanicOrDefault;
 // Rust code, we model these thread locals as a stack, so we can always
 // restore the previous version.
 thread_local! {
-    pub static USERDATA: RefCell<Vec<*mut c_void>> = RefCell::new(Vec::new());
+    pub static USERDATA: RefCell<Vec<Userdata>> = RefCell::new(Vec::new());
+}
+
+pub struct Userdata {
```

Comment:
One might consider simply storing the rustls_connection here. It seems to contain the application context which is exactly what we want to push/pop with this.

If my Rust were better, I'd also knew if a `Arc<dyn AppContextTrait>` could be used here.

Reply:
I considered this, but this would require us to violate the borrowing rules for `Connection` in ways we don't already. Specifically in the case where we are in a call that takes a `*mut rustls_connection` and turns it into a `&mut Connection`, we would then have to put a `*mut Connection` in `Userdata` and use unsafe to dereference it. I think this winds up being sound due to the threading properties of rustls (everything happens on the thread you called on), but it still seems like unnecessary risk.

Of course we are taking the same risk (having multiple mutable pointers) with `userdata: *mut c_void`, but that is unavoidable risk, if we want a general-purpose callback.